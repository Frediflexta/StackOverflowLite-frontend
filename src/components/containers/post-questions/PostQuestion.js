import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Editor } from '@tinymce/tinymce-react';
import propTypes from 'prop-types';
import Button from '../../presentation/button/Button';
import Notification from '../../presentation/notification/Notification';
import { postQuestionRequest } from '../../../redux/actions/postQuestions/postQuestionAction';
import '../../../utils/styles.scss';

class PostQuestion extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      body: '',
      status: '',
      message: '',
      display: 'none',
    };
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  handleEditorChange = (e) => {
    e.target.getContent();
  }

  handleInputChange(event) {
    this.setState({
      [event.target.id]: event.target.value,
    });
  }

  handleQuestionSubmit = async (event) => {
    event.preventDefault();
    const {
      postQues
    } = this.props;
    try {
      await postQues({ ...this.state });
      this.setState({
        display: 'block',
        message: this.props.notification.message,
        status: this.props.notification.type,
      });
      setTimeout(() => {
        this.setState({
          display: 'none',
          message: '',
          status: '',
          title: '',
          body: '',
        });
      }, 3000);
    } catch (error) {
      // do nothing
    }
  }

  render() {
    const {
      status,
      message,
      display
    } = this.state;

    return (
      <div className="columns is-mt-10">
        <div className="column is-4 is-offset-4">
          <h1 className="title is-4">Ask Your Questions</h1>
          <form onSubmit={this.handleQuestionSubmit}>
            <Notification
              status={status}
              message={message}
              display={display}
            />
            <input
              className="input"
              type="text"
              placeholder="Question title"
              id="title"
              value={this.state.title}
              onChange={this.handleInputChange}
            />
            <br /><br />
            {/* <textarea
              className="textarea"
              placeholder="Say something..."
              rows="10"
              id="body"
              value={this.state.body}
              onChange={this.handleInputChange}
            /> */}
            <Editor
              placeholder="Say something..."
              init={{
                plugins: 'link image code',
                toolbar: 'undo redo | bold italic | alignleft aligncenter alignright | code'
              }}
              id="body"
              onChange={this.handleEditorChange}
            />
            <br/> <br/>
            <Button
              className="button is-info"
              type="submit"
              text="Post Question"
            />
          </form>
        </div>
      </div>
    );
  }
}

PostQuestion.propTypes = {
  loading: propTypes.bool,
  postQues: propTypes.func,
  notification: propTypes.object,
};

const mapStateToProps = state => ({
  questionBody: state.PostQuestions,
  notification: state.notificationReducer,
});

const mapDispatchToProps = dispatch => ({
  postQues: data => dispatch(postQuestionRequest(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(PostQuestion);
