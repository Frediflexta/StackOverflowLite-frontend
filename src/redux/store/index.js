import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/rootReducer';

// const middleware = process.env.NODE_ENV !== 'production'
//   ? [require('redux-immutable-state-invariant').default(), thunk]
//   : [thunk];

const middleware = [thunk];

if (process.env.NODE_ENV !== 'production') {
  const { createLogger } = require('redux-logger');
  const logger = createLogger({ collapsed: true });
  middleware.push(require('redux-immutable-state-invariant').default());
  middleware.push(logger);
}

/**
 * @param {*} initialState
 * @return {*} createStore
 */
function configureStore() {
  return createStore(
    rootReducer,
    applyMiddleware(...middleware),
  );
}

const store = configureStore();
window.store = store;

store.subscribe(() => localStorage.setItem('store', JSON.stringify(store.getState())));
export default store;
