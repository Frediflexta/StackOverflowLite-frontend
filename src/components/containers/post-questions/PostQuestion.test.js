import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import PostQuestion from './PostQuestion';
import store from '../../../redux/store/index';

describe('<PostQuestion /> rendering tests', () => {
  const props = {
    handleQuestionSubmit: jest.fn(),
    questions: [
      { title: 'test question' },
      { body: 'test body' },
    ],
    notification: { type: 'success' },
  };
  let wrapper;

  beforeEach(() => {
    jest.setTimeout(10000);
    wrapper = mount(
      <MemoryRouter>
        <PostQuestion {...props} store={store}/>
      </MemoryRouter>
    );
  });

  it('should render PostQuestion successfully', () => {
    expect(wrapper.find('input').exists()).toBe(true);
  });

  it('matches the snapshot', () => {
    const tree = shallow(<PostQuestion {...props} store={store}/>);
    expect(tree).toMatchSnapshot();
  });

  // it('should handle input changes successfully', () => {
  //   const instance = wrapper.find('PostQuestion').instance();

  //   const ques = {
  //     title: 'test',
  //     body: 'email'
  //   };

  //   instance.handleInputChange(ques);
  //   expect(instance.state.title).toEqual('test');
  // });

  it('should handle submit question form successfully', () => {
    const instance = wrapper.find('PostQuestion').instance();
    const event = {
      preventDefault: jest.fn()
    };

    instance.props.handleQuestionSubmit(event);
  });
});
