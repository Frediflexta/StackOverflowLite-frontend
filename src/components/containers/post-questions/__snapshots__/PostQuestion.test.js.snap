// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostQuestion /> rendering tests matches the snapshot 1`] = `
<PostQuestion
  handleQuestionSubmit={[MockFunction]}
  notification={
    Object {
      "message": null,
      "type": null,
    }
  }
  postQues={[Function]}
  questionBody={
    Object {
      "data": Object {},
      "error": false,
      "loading": false,
      "success": false,
    }
  }
  questions={
    Array [
      Object {
        "title": "test question",
      },
      Object {
        "body": "test body",
      },
    ]
  }
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
